File: challenge/leet.gcc/easy/q661/q661.cpp
/**
 * 661. Image Smoother
 * https://leetcode.com/problems/image-smoother/description/ 
 */

#include <vector>
#include <iostream>
#include <numeric>
#include <string>

using namespace std;

class Solution {
public:
    void reducerow(vector<int>& v, vector<int>& buf) {
        if (v.size() > 1) {
            buf[0] = v[0] + v[1];

            for (int i = 1; i < v.size() - 1; i++)
                buf[i] = v[i - 1] + v[i] + v[i + 1];

            buf[v.size() - 1] = v[v.size() - 2] + v[v.size() - 1];
        }
        else buf[0] = v[0];
    }

    vector<vector<int>> imageSmoother(vector<vector<int>>& img) {
        int m = img.size(), n = img[0].size();
        vector<vector<int>> buf;

        for (int r = 0; r < m; r++) {
            vector<int> brow(n);
            reducerow(img[r], brow);
            buf.push_back(brow);
        }

        int d0 = max(m, n) == 1 ? 1 : min(m, n) == 1 ? 2 : 4;
        int dx = min(m, n) == 1 ? 3 : 6;
        int di = 9;

        for (int c = 0; c < n; c++) {
            // collectcol(c, buf);
            img[0][c] = (m > 1 ? buf[0][c] + buf[1][c] : buf[0][c]) / (c == 0 or c == n-1 ? d0 : dx);

            for (int r = 1; r < m - 1; r++) {
                img[r][c] = (buf[r - 1][c] + buf[r][c] + buf[r + 1][c]) / (c == 0 or c == n-1 ? dx : di);
            }
            if (m > 1)
                img[m - 1][c] = (buf[m - 2][c] + buf[m - 1][c]) / (c == 0 or c == n-1 ? d0 : dx);
        }

        return img;
    }
};

string reduce(vector<int> v) {
    return reduce(next(v.begin()), v.end(), string("[") + to_string(v[0]),
        [](string a, int b) { return move(a) + "," + to_string(b); }) + string("]");
}

string reduce(vector<vector<int>> v) {
    return reduce(next(v.begin()), v.end(), string("[") + reduce(v[0]),
        [](string a, vector<int> b) { return move(a) + "," + reduce(b); }) + string("]");
}

int main() {
    Solution s;
    vector<vector<int>> v;

    v = vector<vector<int>>{ vector<int>{ 1,2 }, vector<int>{2,3}, vector<int>{4,5} };
    cout << "[[2,2],[2,2],[3,3]] ?" << endl << reduce(s.imageSmoother(v)) << endl;
}

File: challenge/leet/easy/q661.py
'''
 * 661. Image Smoother
 * https://leetcode.com/problems/image-smoother/description/
'''
import unittest
from typing import List


class Solution:
    def imageSmoother(self, img: List[List[int]]) -> List[List[int]]:
        m, n = len(img), len(img[0])
        buf = [[0] * n for _ in range(m)]

        if n > 1:
            for rx, r in enumerate(img):
                buf[rx][0] = r[0] + r[1]
                if n > 2:
                    for cx in range(1, n - 1):
                        buf[rx][cx] = r[cx - 1] + r[cx] + r[cx + 1]
                buf[rx][-1] = r[-2] + r[-1]
        else:
            img, buf = buf, img

        d0 = 1 if max(m, n) == 1 else 2 if min(m, n) == 1 else 4
        dx = 6 if min(m, n) > 1 else 3
        di = 9
        if m > 1:
            for cx in range(n):
                img[0][cx] = (buf[0][cx] + buf[1][cx]) // (dx if 0 < cx < n - 1 else d0)
                # if m > 2:
                for rx in range(1, m - 1):
                    img[rx][cx] = (buf[rx - 1][cx] + buf[rx][cx] + buf[rx + 1][cx]) // (di if 0 < cx < n - 1 else dx)
                else:
                    img[-1][cx] = (buf[-2][cx] + buf[-1][cx]) // (dx if 0 < cx < n - 1 else d0)
        else:  # m == 1
            for cx in range(n):
                img[0][cx] = buf[0][cx] // (dx if 0 < cx < n - 1 else d0)

        return img


class MyTestCase(unittest.TestCase):
    def test_something(self):
        s = Solution()

        self.assertEqual([[3,3]], s.imageSmoother([[1,5]]))
        self.assertEqual([[137, 93, 78, 55], [141, 97, 87, 55], [137, 93, 93, 78]],
             s.imageSmoother([[100, 200, 10, 10], [200, 50, 2, 200], [100, 200, 11, 100]]))
        self.assertEqual([[137, 141, 137], [141, 138, 141], [150, 144, 147], [162, 142, 158]],
             s.imageSmoother([[100, 200, 100], [200, 50, 200], [100, 200, 100], [120, 230, 103]]))
        self.assertEqual([[1, 1, 1]], s.imageSmoother([[1, 1, 1]]))
        self.assertEqual([[0, 0, 0], [0, 0, 0], [0, 0, 0]], s.imageSmoother([[1, 1, 1], [1, 0, 1], [1, 1, 1]]))
        self.assertEqual([[137, 141, 137], [141, 138, 141], [137, 141, 137]],
             s.imageSmoother([[100, 200, 100], [200, 50, 200], [100, 200, 100]]))
        self.assertEqual([[1], [1]], s.imageSmoother([[1], [1]]))


if __name__ == '__main__':
    unittest.main()
